packetbeat/protos/cassandra/internal/gocql/frame.go

```diff
@@ -0,0 +1,530 @@
+
+	flag, err := f.r.ReadByte()
+	if err != nil {
+		return nil, err
+	}
+	head.Flags = flag
+
+	if version > protoVersion2 {
+		stream, err := f.r.ReadNetUint16()
+		if err != nil {
+			return nil, err
+		}
+		head.Stream = int(stream)
+
+		b, err := f.r.ReadByte()
+		if err != nil {
+			return nil, err
+		}
+
+		head.Op = FrameOp(b)
+		l, err := f.r.ReadNetUint32()
+		if err != nil {
+			return nil, err
+		}
+		head.BodyLength = int(l)
+	} else {
+		stream, err := f.r.ReadNetUint8()
+		if err != nil {
+			return nil, err
+		}
+		head.Stream = int(stream)
+
+		b, err := f.r.ReadByte()
+		if err != nil {
+			return nil, err
+		}
+
+		head.Op = FrameOp(b)
+		l, err := f.r.ReadNetUint32()
+		if err != nil {
+			return nil, err
+		}
+		head.BodyLength = int(l)
+	}
+
+	if head.BodyLength < 0 {
+		return nil, fmt.Errorf("frame body length can not be less than 0: %d", head.BodyLength)
+	} else if head.BodyLength > maxFrameSize {
+		// need to free up the connection to be used again
+		logp.Err("head length is too large")
+		return nil, ErrFrameTooBig
+	}
+
+	headSize := f.r.BufferConsumed()
+	head.HeadLength = headSize
+
+	debugf("header: %v", head)
+
+	f.Header = head
+	return head, nil
+}
+
+// reads a frame form the wire into the framers buffer
+func (f *Framer) ReadFrame() (data map[string]interface{}, err error) {
+
+	defer func() {
+		if r := recover(); r != nil {
+			if _, ok := r.(runtime.Error); ok {
+				panic(r)
+			}
+			err = r.(error)
+		}
+	}()
+
+	decoder := &StreamDecoder{}
+	decoder.r = f.r
+	f.decoder = decoder
+
+	data = make(map[string]interface{})
+
+	//Only QUERY, PREPARE and EXECUTE queries support tracing
+	//If a response frame has the tracing flag set, its body contains
+	//a tracing ID. The tracing ID is a [uuid] and is the first thing in
+	//the frame body. The rest of the body will then be the usual body
+	//corresponding to the response opcode.
+	if f.Header.Flags&flagTracing == flagTracing && (f.Header.Op&opQuery == opQuery || f.Header.Op&opExecute == opExecute || f.Header.Op&opPrepare == opPrepare) {
+
+		debugf("tracing enabled")
+
+		//seems no UUID to read, protocol incorrect?
+		//uid := decoder.ReadUUID()
+		//data["trace_id"] = uid.String()
+	}
+
+	if f.Header.Flags&flagWarning == flagWarning {
+		debugf("hit warning flags")
+
+		warnings := decoder.ReadStringList()
+		// dealing with warnings
+		data["warnings"] = warnings
+	}
+
+	if f.Header.Flags&flagCustomPayload == flagCustomPayload {
+		debugf("hit custom payload flags")
+
+		f.Header.CustomPayload = decoder.ReadBytesMap()
+	}
+
+	if f.Header.Flags&flagCompress == flagCompress {
+		//decompress data and switch to use bytearray decoder
+		decoder := &ByteArrayDecoder{}
+		buf := make([]byte, f.Header.BodyLength)
+		f.r.Read(buf)
+		dec, err := f.compres.Decode(buf)
+		if err != nil {
+			return nil, err
+		}
+
+		decoder.Data = &dec
+		f.decoder = decoder
+
+		debugf("hit compress flags")
+
+		return nil, errors.New("Compressed content not supported yet")
```

Comment:
so we first decompress body and return error about not supporting compressed content next? I'm a little confused what's correct here.